### DISCUSSION TITLE
Feature idea: Automatically fix issues

### DISCUSSION THREAD

**Discussion by tbarn** [stoplightio/spectral#98] [Ideas]
Would love to open it for discussion on what the preferred expected behavior would be. 

I think this could be more of a CLI feature, so then you can interactively accept and reject changes to a document. Also, you could be prompted to fill out things. For example, in an OpenAPI document, descriptions are often forgotten. The CLI could prompt you to fix them out.  

### What is the motivation / use case for changing the behavior?
Fix errors and warnings in a document easier and faster

**Comment by sebas2day** [comment]
Is this something that has been considered on actually implementing within Spectral? 

If so, are there any ideas already on how this should get implemented? Would it be something similar to [eslint auto fixing](https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes)?

Next, what auto fixable rules can we already think of?
 - openapi-tags-alphabetical: auto re-order
 - contact-properties: scaffold some sample snippet
 - oas3-server-trailing-slash: auto remove trailing slash
 - oas3-unused-components-schema: auto remove (can be tricky / undesired)
 - ... others?

I might be willing to put time on implementing this since I see a great benefit on having this. I would love to have it extensible as well so I can integrate custom auto fix functionality with custom defined rules.

Let me know your ideas.

**Reply by fjaeger** [reply]
`operationId` could also easily be "auto-fixed" by combining some other attributes, such as `tags` and the operation itself

**Reply by thim81** [reply]
Hi @sebas2day 
About the sorting, I created [openapi-format ](https://www.npmjs.com/package/openapi-format) explicitly for that purpose.
It is a small CLI that helps you to keep your OpenAPI document in a good shape, by doing [ordering](https://www.npmjs.com/package/openapi-format#openapi-sort-configuration-options) & [formatting](https://www.npmjs.com/package/openapi-format#openapi-formatting-configuration-options).

I use spectral to validate if the OpenAPI document meets the required structure and content validation and openapi-format to handle the ordering & formatting. So maybe that way of working is useful for you, hence that I'm sharing my approach.
